### PR TITLE
Syntax Highlighting for Diff

### PR DESCRIPTION
**Pull request by TheWirtschaftsmann**
to #443

Hi @larshp, 
Here a small contribution on syntax highlighting. Doesn't work perfectly, there is still area for improvement, but at least it works and looks nicely:)

image

### CODE REVIEW

**Comment by sbcgua** [comment]
@larshp  seems like some issue with linter ...

P.S. a comment from my side. The regex with keywords compiles quite long - ~2 sec. So the first diff takes a bit long to display. If this will be annoying we might need to reduce the number of keywords a bit (I guess there are a lot of rarely used ;)

**Comment by larshp** [comment]
cool, can you replace line 346 with `substring(` instead? http://help.sap.com/abapdocu_702/en/abensubstring_functions.htm
it should fix the error from abaplint, plus make it a bit more readable

I've created https://github.com/larshp/abaplint/issues/114 but not sure I have time to fix it this week

**Comment by TheWirtschaftsmann** [comment]
Hi Lars, 

Thank you for suggestion regards substring() function. It is indeed more efficient than the previous version like lv_chunk = iv_line+<match>-offset(<match>-length). By the way, I've found another bug like that in the source code of highlighter that I fixed along the way.

File: src/zabapgit_syntax_highlighter.prog.abap [hunk 1]
```diff
@@ -0,0 +1,762 @@
...
+
+    DATA: lt_matches TYPE ty_match_tt.
+
+    IF strlen( iv_line ) = 0.
+      RETURN.
+    ENDIF.
+
+    lt_matches = me->parse_line( iv_line ).
+
+    me->order_matches( EXPORTING iv_line    = iv_line
```
Comment (EduardoCopat) [inline src/zabapgit_syntax_highlighter.prog.abap]: Why is it needed to order matches?

File: src/zabapgit_syntax_highlighter.prog.abap [hunk 2]
```diff
@@ -0,0 +1,762 @@
...
+
+ENDCLASS.                       " lcl_code_highlighter IMPLEMENTATION
+
+
+*----------------------------------------------------------------------*
+*       CLASS ltcl_code_highlighter definition
+*----------------------------------------------------------------------*
+CLASS ltcl_code_highlighter DEFINITION FINAL
+  FOR TESTING RISK LEVEL HARMLESS DURATION SHORT
+  INHERITING FROM cl_aunit_assert.
```
Comment (EduardoCopat) [inline src/zabapgit_syntax_highlighter.prog.abap]: inheriting from cl_aunit_assert (see class documentatino) is deprecated, use cl_abap_unit_assert

File: src/zabapgit_syntax_highlighter.prog.abap [hunk 3]
```diff
@@ -0,0 +1,762 @@
...
+
+  PRIVATE SECTION.
+
+    DATA:
+          mo  TYPE REF TO lcl_code_highlighter.
```
Comment (EduardoCopat) [inline src/zabapgit_syntax_highlighter.prog.abap]: Could it be renamed to mo_code_highlighter?

File: src/zabapgit_syntax_highlighter.prog.abap [hunk 4]
```diff
@@ -0,0 +1,762 @@
...
+    ELSEIF &5 = 'ORDER'.
+      APPEND ls_match to lt_after_order.
+    ENDIF.
+  END-OF-DEFINITION.
+
+  METHOD setup.
+    CREATE OBJECT mo.
+  ENDMETHOD.                    " setup
+
+  METHOD parse_and_order.
```
Comment (EduardoCopat) [inline src/zabapgit_syntax_highlighter.prog.abap]: This test method contains several different logical assertions. Should it be split into others?

File: src/zabapgit_syntax_highlighter.prog.abap (hunk 3 above)
Comment (sbcgua) [inline src/zabapgit_syntax_highlighter.prog.abap]: too long name for a test class - this is the object under test.  ;)

File: src/zabapgit_syntax_highlighter.prog.abap (hunk 1 above)
Comment (TheWirtschaftsmann) [inline src/zabapgit_syntax_highlighter.prog.abap]: Hi @EduardoCopat ,
Shortly the idea behind the solution is as follows: 
- parse_line() - returns the table with all matches for texts, key words and comments.
- order_matches() - processes this table - removes some matches, which are not necessary (e.g. key words inside code that is commented out), fine-tunes the values of some matches and adds additional lines which represent plain code, which should not be formatted. At the end, each line in ct_matches represent exact part of the string and details necessary for formatting (starting position + length + reference to style);
- format_line() is quite simple - processes the table entry by entry and passes it to apply_style(). 

Hope that answers your question.

File: src/zabapgit_syntax_highlighter.prog.abap (hunk 4 above)
Comment (TheWirtschaftsmann) [inline src/zabapgit_syntax_highlighter.prog.abap]: Well... at first it was devised as two separate methods for testing - for parsing and ordering. But as it takes a lot of code to generate tables with matches, I decided to combine them into one in order to improve the readability. Thus, I generate in one shot expected values for parsing and for ordering and then test both methods. Examples are rather unclear if they are separated.

File: src/zabapgit_syntax_highlighter.prog.abap (hunk 2 above)
Comment (TheWirtschaftsmann) [inline src/zabapgit_syntax_highlighter.prog.abap]: Fair enough. I'll work on it and will submit it a bit later, probably in the evening.

File: src/zabapgit_syntax_highlighter.prog.abap (hunk 1 above)
Comment (EduardoCopat) [inline src/zabapgit_syntax_highlighter.prog.abap]: Yes, now it is clear. Thanks!

File: src/zabapgit_syntax_highlighter.prog.abap (hunk 3 above)
Comment (EduardoCopat) [inline src/zabapgit_syntax_highlighter.prog.abap]: @sbcgua Even though it is larger, I prefer that in order to improve readability. Example:

`mo->process_line()`

`mo_code_highlighter->process_line()`

Second one, in my opinion, is more readable and easier to understand.

File: src/zabapgit_syntax_highlighter.prog.abap (hunk 4 above)
Comment (EduardoCopat) [inline src/zabapgit_syntax_highlighter.prog.abap]: Got it. Maybe it's a matter of preference. I prefer to isolate each test case in a separate test method. more on the why [here](http://softwareengineering.stackexchange.com/questions/7823/is-it-ok-to-have-multiple-asserts-in-a-single-unit-test)

File: src/zabapgit_syntax_highlighter.prog.abap (hunk 2 above)
Comment (TheWirtschaftsmann) [inline src/zabapgit_syntax_highlighter.prog.abap]: Change to cl_abap_unit_assert is delivered in my latest commit "Refactoring of UT". Hope that's the way it should be:)

**Comment by sbcgua** [comment]
OK. Merging then. :)
In case the diff page start will take long and it will be annoying please open an issue - we'll look into it.

**Pull request merged**